File: src/crypto/encoding.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 The MCDF Project
#include "mcdf/crypto/encoding.hpp"

#include <cstdint>
#include <vector>

namespace mcdf {
namespace {

constexpr char kB64[] =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
constexpr char kB58[] =
    "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

int b64_value(char c) {
  if (c >= 'A' && c <= 'Z') return c - 'A';
  if (c >= 'a' && c <= 'z') return c - 'a' + 26;
  if (c >= '0' && c <= '9') return c - '0' + 52;
  if (c == '-') return 62;
  if (c == '_') return 63;
  return -1;
}

int b58_value(char c) {
  for (int i = 0; i < 58; ++i)
    if (kB58[i] == c) return i;
  return -1;
}

}  // namespace

std::string base64url_encode(std::string_view data) {
  std::string out;
  out.reserve((data.size() + 2) / 3 * 4);
  std::size_t i = 0;
  const auto byte = [&](std::size_t k) {
    return static_cast<std::uint32_t>(static_cast<unsigned char>(data[k]));
  };
  for (; i + 3 <= data.size(); i += 3) {
    const std::uint32_t n = (byte(i) << 16) | (byte(i + 1) << 8) | byte(i + 2);
    out.push_back(kB64[(n >> 18) & 63]);
    out.push_back(kB64[(n >> 12) & 63]);
    out.push_back(kB64[(n >> 6) & 63]);
    out.push_back(kB64[n & 63]);
  }
  const std::size_t rem = data.size() - i;
  if (rem == 1) {
    const std::uint32_t n = byte(i) << 16;
    out.push_back(kB64[(n >> 18) & 63]);
    out.push_back(kB64[(n >> 12) & 63]);
  } else if (rem == 2) {
    const std::uint32_t n = (byte(i) << 16) | (byte(i + 1) << 8);
    out.push_back(kB64[(n >> 18) & 63]);
    out.push_back(kB64[(n >> 12) & 63]);
    out.push_back(kB64[(n >> 6) & 63]);
  }
  return out;
}

Result<std::string> base64url_decode(std::string_view text) {
  std::string out;
  std::uint32_t buf = 0;
  int bits = 0;
  for (const char c : text) {
    const int v = b64_value(c);
    if (v < 0) return fail(ErrorCode::kParse, "invalid base64url character");
    buf = (buf << 6) | static_cast<std::uint32_t>(v);
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      out.push_back(static_cast<char>((buf >> bits) & 0xFF));
    }
  }
  return out;
}

std::string base58btc_encode(std::string_view data) {
  std::size_t zeros = 0;
  while (zeros < data.size() &&
         static_cast<unsigned char>(data[zeros]) == 0) {
    ++zeros;
  }
  std::vector<std::uint8_t> digits;  // little-endian base-58
  for (std::size_t i = zeros; i < data.size(); ++i) {
    int carry = static_cast<unsigned char>(data[i]);
    for (auto& d : digits) {
      const int x = d * 256 + carry;
      d = static_cast<std::uint8_t>(x % 58);
      carry = x / 58;
    }
    while (carry) {
      digits.push_back(static_cast<std::uint8_t>(carry % 58));
      carry /= 58;
    }
  }
  std::string out(zeros, '1');
  for (auto it = digits.rbegin(); it != digits.rend(); ++it)
    out.push_back(kB58[*it]);
  return out;
}

Result<std::string> base58btc_decode(std::string_view text) {
  std::size_t zeros = 0;
  while (zeros < text.size() && text[zeros] == '1') ++zeros;
  std::vector<std::uint8_t> bytes;  // little-endian base-256
  for (std::size_t i = zeros; i < text.size(); ++i) {
    const int v = b58_value(text[i]);
    if (v < 0) return fail(ErrorCode::kParse, "invalid base58 character");
    int carry = v;
    for (auto& b : bytes) {
      const int x = b * 58 + carry;
      b = static_cast<std::uint8_t>(x % 256);
      carry = x / 256;
    }
    while (carry) {
      bytes.push_back(static_cast<std::uint8_t>(carry % 256));
      carry /= 256;
    }
  }
  std::string out(zeros, '\0');
  for (auto it = bytes.rbegin(); it != bytes.rend(); ++it)
    out.push_back(static_cast<char>(*it));
  return out;
}

}  // namespace mcdf

File: apps/web/packages/mcdf-ts/src/crypto/encoding.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 Sukesh Ashok Kumar and The MCDF Project
//
// base64url (RFC 4648 §5, unpadded) and base58btc — the two alphabets MCDF
// needs: JWS segments are base64url, `did:key` is base58btc after the `z`
// multibase prefix.
//
// Written in-house rather than taken from a package. Both are
// tiny, both are on the byte-parity path with `src/crypto/encoding.cpp`, and a
// dependency whose output must match another implementation character for
// character is a dependency worth not having.

export class EncodingError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'EncodingError';
  }
}

const B64 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';
const B58 = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

/** Reverse lookup tables, built once — the decoders run per signature segment. */
const B64_VALUES = new Int8Array(128).fill(-1);
for (let i = 0; i < B64.length; i++) B64_VALUES[B64.charCodeAt(i)] = i;

const B58_VALUES = new Int8Array(128).fill(-1);
for (let i = 0; i < B58.length; i++) B58_VALUES[B58.charCodeAt(i)] = i;

export function base64urlEncode(data: Uint8Array): string {
  let out = '';
  let i = 0;
  for (; i + 3 <= data.length; i += 3) {
    const n = ((data[i] as number) << 16) | ((data[i + 1] as number) << 8) | (data[i + 2] as number);
    out += B64[(n >> 18) & 63];
    out += B64[(n >> 12) & 63];
    out += B64[(n >> 6) & 63];
    out += B64[n & 63];
  }
  const rest = data.length - i;
  if (rest === 1) {
    const n = (data[i] as number) << 16;
    out += B64[(n >> 18) & 63];
    out += B64[(n >> 12) & 63];
  } else if (rest === 2) {
    const n = ((data[i] as number) << 16) | ((data[i + 1] as number) << 8);
    out += B64[(n >> 18) & 63];
    out += B64[(n >> 12) & 63];
    out += B64[(n >> 6) & 63];
  }
  return out;
}

/**
 * Decodes unpadded base64url. `=` padding is tolerated on input (some producers
 * emit it even where JWS forbids it) but never produced.
 */
export function base64urlDecode(text: string): Uint8Array {
  const out: number[] = [];
  let buffer = 0;
  let bits = 0;
  for (const ch of text) {
    if (ch === '=') continue;
    const code = ch.charCodeAt(0);
    const value = code < 128 ? (B64_VALUES[code] as number) : -1;
    if (value < 0) throw new EncodingError(`invalid base64url character: ${ch}`);
    buffer = (buffer << 6) | value;
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      out.push((buffer >> bits) & 0xff);
    }
  }
  return Uint8Array.from(out);
}

export function base58btcEncode(data: Uint8Array): string {
  let zeros = 0;
  while (zeros < data.length && data[zeros] === 0) zeros++;

  const digits: number[] = []; // little-endian base-58
  for (let i = zeros; i < data.length; i++) {
    let carry = data[i] as number;
    for (let d = 0; d < digits.length; d++) {
      const x = (digits[d] as number) * 256 + carry;
      digits[d] = x % 58;
      carry = (x / 58) | 0;
    }
    while (carry > 0) {
      digits.push(carry % 58);
      carry = (carry / 58) | 0;
    }
  }

  let out = '1'.repeat(zeros);
  for (let i = digits.length - 1; i >= 0; i--) out += B58[digits[i] as number];
  return out;
}

export function base58btcDecode(text: string): Uint8Array {
  let zeros = 0;
  while (zeros < text.length && text[zeros] === '1') zeros++;

  const bytes: number[] = []; // little-endian base-256
  for (let i = zeros; i < text.length; i++) {
    const code = text.charCodeAt(i);
    const value = code < 128 ? (B58_VALUES[code] as number) : -1;
    if (value < 0) throw new EncodingError(`invalid base58 character: ${text[i]}`);
    let carry = value;
    for (let b = 0; b < bytes.length; b++) {
      const x = (bytes[b] as number) * 58 + carry;
      bytes[b] = x % 256;
      carry = (x / 256) | 0;
    }
    while (carry > 0) {
      bytes.push(carry % 256);
      carry = (carry / 256) | 0;
    }
  }

  const out = new Uint8Array(zeros + bytes.length);
  for (let i = 0; i < bytes.length; i++) out[zeros + i] = bytes[bytes.length - 1 - i] as number;
  return out;
}
